org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatFilesHandler.java

```diff
@@ -81,15 +81,21 @@ protected void formatFile(final @NonNull IFile file, final IProgressMonitor moni
 				return;
 
 			monitor.setTaskName(NLS.bind(Messages.LSPFormatFilesHandler_FormattingFile, file.getFullPath()));
-			final VersionedFormatRequest formatRequest = formatter.versionedRequestFormatting(doc,
-					new TextSelection(0, 0));
+			final var formatting = formatter.requestFormatting(doc,
```

Comment:
My guess for the NPEs is, that the formatter does not wait correctly for an initialized LS. which may not be an issue if you do formatting of an open text document (where the LS is already loaded) but in case if this batch formatting handler, there might not be any LS running already. That is why in #398 I tried to run the formatter within the `computeFirst` callback.

Reply:
> I gave your branch a try, two observations:
> 
> a) batch formatting became extremely slow, like one second per file. b) this error was logged a lot during formatting:
> 
> ```
> an 25, 2023 1:39:42 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
> SEVERE: Internal error: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4j.ServerCapabilities.getCodeActionProvider()" because "this.serverCapabilities" is null
> java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4j.ServerCapabilities.getCodeActionProvider()" because "this.serverCapabilities" is null
> 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
> 	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
> 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
> 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
> 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
> 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
> Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.lsp4j.ServerCapabilities.getCodeActionProvider()" because "this.serverCapabilities" is null
> 	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$16(LanguageServerWrapper.java:928)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
> 	at org.eclipse.lsp4e.LanguageServerWrapper.registerCapability(LanguageServerWrapper.java:881)
> 	at org.eclipse.lsp4e.LanguageClientImpl.lambda$2(LanguageClientImpl.java:136)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
> 	... 6 more
> ```
Thanks - I'll take a look. Can you give me any pointers as to how to use this class so I can try and reproduce the problems?

Reply:
just right click on a source folder of a project that uses a LS and select "source/format"

Reply:
> My guess for the NPEs is, that the formatter does not wait correctly for an initialized LS. which may not be an issue if you do formatting of an open text document (where the LS is already loaded) but in case if this batch formatting handler, there might not be any LS running already. 

Hmm both new and old API are supposed to have the property that actions should wait for the running LS - that property, in conjunction with the fact we mostly lazily start LS on demand accounts for a lot of the complexity!

The pseudocode for how LSP4e mostly works is `getRunningLS().thenDoStuff()`.

Reply:
It was just a wild guess, but I am not seeing these exception with #398.

Reply:
Are you running batch processing with a very short idle timeout and no documents open in the IDE? LSP4e is centred around documents open in the IDE - the lifecycle for the LS times it out and shuts it down (temporarily - it can restart) a certain time after the last open document attached the LS has closed.

This could explain both the timeout and the errors: if you run a batch command on a document without it being open in the IDE then the LS could get shut down immediately after the operation. If the LS gets lazily started then shut down per document in batch then that would explain the terrible throughput.

Reply:
The code for LSPDiagnosticsToMarkers has special logic related to this kind of thing IIRC - when LS is launched for a project (triggered by a document opening) the LS can scan the whole project and send back diagnostics for files not open in the IDE, so LSP4e opens the document temporarily to create the markers.

Reply:
Thanks, I've tried running using Rust as a LS formatting a project without any files open in the IDE and can see a lot of this sort of problem and symptoms of the server starting up and shutting down all the time. I think this is a more general problem with batch processing and the idle timeout - there's nothing to keep the server up. I'll check the behaviour on master to see how it looks there, but I'd be surprised if it wasn't an issue without my changes (I may have moved the symptoms).

If you keep open a file from within the project you are batch-formatting does it behave better?

Reply:
yes if I have one file open it is better or if I use the apporach of #398